posts/tools/test-frame-introduction.md

```diff
@@ -0,0 +1,174 @@
+direct communication with Kubernetes API.
+Because of this library (and also many others like [JUnit5](https://junit.org/junit5/)),
+we decided to write the test suite in Java, which is also the language that our
+application is written in.
+
+## Creating a test-suite
+
+In the beginning, you start with observing what is needed during each phase of the test case.
+You can start with the setup needed before the test execution.
+Then, during the test case, you can create and delete the Kubernetes resources, do operations
+that your application allows, check logs of the Pod, validate results, and clean-up the environment after
+the test case.
+
+Having a clean-up phase in every test case is sufficient when you have just a few test cases.
+But once you are writing more and more tests, it starts to be boring and repetitive when you
+have to clean the resources at the end of each test.
+
+To make your life easier, you create a mechanism that will store the resources created in the test
+and once the test is completed, it will clear everything based on the LIFO mechanism.
+
+Great, what about the situation when the test fails?
+You would have to simulate the same environment and situation in which it failed.
+But sometimes, it is not that simple to reproduce the exact behavior.
+Because of this, you create another mechanism for collecting logs, description of resources, and many more,
+that will be executed in case of the test failure.
+
+All of these will be mostly stored in the repository where you are writing the tests, which makes sense
+and it is the best possible solution at that moment.
+
+## More repositories, more redundancy
+
+In case we have just one repository where all the testing is done, everything is fine.
+But once we want to extend the product and add new components that will also need system testing,
+we start a fight with redundancy.
+And, one day, you will find yourself copying the testing-related classes across multiple repositories,
+mixing things and always trying to refactor the code without success and without "the light at the end of the tunnel" that
+there will be an end to this.
+At least, that is the situation in which we ended up.
+
+Yes, you can use the test classes from the main repository as a Maven dependency, in case you are releasing it together
+with other modules to the Maven Central (or anywhere else).
+However, there are situations when you cannot do that - one of the examples is cyclic dependency.
+
+Let's say that you have the main repository, containing the code for an operator, and you have a few other repositories for the components that
+are managed by the operator.
+Each of the components has a different release cadence, so having them in the operator's repository is not the correct approach.
+In this particular case, if you would use the "system test" dependency from the operator's repository inside the component's repository (and the component
```

Comment:
Wouldn't be better to use `system test` instead of "" ?

Reply:
removed "", but it's not a name of module or something, so i didn't use the ``